funsor/minipyro.py

```diff
@@ -408,24 +400,33 @@ def __init__(self, model, guide, optim, loss):
 
     # This method handles running the model and guide, constructing the loss
     # function, and taking a gradient step.
-    def step(self, *args, **kwargs):
+    def run(self, *args, **kwargs):
         # This wraps both the call to `model` and `guide` in a `trace` so that
         # we can record all the parameters that are encountered. Note that
         # further tracing occurs inside of `loss`.
-        with trace() as param_capture:
-            # We use block here to allow tracing to record parameters only.
-            with block(hide_fn=lambda msg: msg["type"] != "param"):
-                loss = self.loss(self.model, self.guide, *args, **kwargs)
-        # Differentiate the loss.
-        funsor.to_data(loss).backward()
-        # Grab all the parameters from the trace.
-        params = [site["value"].data.unconstrained() for site in param_capture.values()]
-        # Take a step w.r.t. each parameter in params.
-        self.optim(params)
-        # Zero out the gradients so that they don't accumulate.
-        for p in params:
-            p.grad = torch.zeros_like(p.grad)
-        return loss.item()
+        with funsor.terms.lazy:
+            with trace() as param_capture:
+                # We use block here to allow tracing to record parameters only.
+                with block(hide_fn=lambda msg: msg["type"] != "param"):
+                    loss = self.loss(self.model, self.guide, *args, **kwargs)
+        init_params = {
+            name: funsor.to_funsor(site["fn"](*site["args"]).data.unconstrained())
+            for name, site in param_capture.items()
+        }
+        with self.optim.with_init(init_params):
+            with funsor.montecarlo.MonteCarlo():
+                result = loss.reduce(funsor.ops.min)
+        for name, value in self.optim.params.items():
+            name = name.replace("_unconstrained", "")
+            value = value.data
+            old_value, constraint = PARAM_STORE[name]
+            if old_value is not value:
+                old_value.data.copy_(value)
+        return result
+
+    def step(self, *args, **kwargs):
+        self.optim.num_steps = 1
+        return self.run(*args, **kwargs)
```

Comment:
for compatibility with `SVI` interface

Reply:
Hmm, let's think about alternative workarounds... One issue here is that the Adam optimizer statistics would not be persisted across svi steps.

One option is simply to change pyroapi's SVI interface to look for either `.run()` or if missing fall back to `.step()`.